dpcpp/test/base/executor.dp.cpp

```diff
@@ -269,4 +269,31 @@ TEST_F(DpcppExecutor, Synchronizes)
 }
 
 
+#define GTEST_ASSERT_NO_EXIT(statement) \
+    ASSERT_EXIT({ {statement} exit(0); }, ::testing::ExitedWithCode(0), "")
+
+
+TEST_F(DpcppExecutor, FreeAfterKernel)
+{
+    testing::FLAGS_gtest_death_test_style = "threadsafe";
+    GTEST_ASSERT_NO_EXIT({
+        size_t length = 10000;
+        auto dpcpp =
+            gko::DpcppExecutor::create(0, gko::ReferenceExecutor::create());
+        {
+            gko::Array<float> x(dpcpp, length);
+            gko::Array<float> y(dpcpp, length);
+            auto x_val = x.get_data();
+            auto y_val = y.get_data();
+            dpcpp->get_queue()->submit([&](sycl::handler &cgh) {
+                cgh.parallel_for(sycl::range<1>{length},
+                                 [=](sycl::id<1> i) { y_val[i] += x_val[i]; });
+            });
+        }
+        // to ensure everything on queue is finished.
+        dpcpp->synchronize();
```

Comment:
Is this necessary? The case we want to catch is the one without synchronization, isn't it?

Reply:
Not sure honestly, I do not know whether queue wait the all kernel finished when queue destroy.
If it is like cuda, just terminate. It is possible that we only have allocation and free but no execution.
I use curly brackets to contain the Array allocation to make sure the free is called before the synchronize.
Yes, it catches the free is not after the kernel finishes

Reply:
When the array gets freed, we wait on the queue, that isn't sufficient?

Reply:
Yes, it is fixed version. when the free does not contain synchronization, we will probably not wait the kernel.

Reply:
Can you remove this line to make sure the program can actually fail due to use-after-free issues?

Reply:
it's shown in the https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/1415200944
and the code contains sync https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/commit/f9da28bc5b8f6847b4d587b0283b2cc7f7e2f60a
I mean fixed version is the we wait on the queue when we call free.
(the following description is from cuda thinking)
Without sync before program end
1. when free does not contain wait:
submit job (not run) -> free -> program end (the job may not run, so we may does not face the free after use)
2. when free does contain wait: (fixed version)
submit job (not run) -> run job (due to free contain wait) -> free -> program end
this case does not need sync in the end

Reply:
I see, thanks! I completely missed that the arrays have their own scope, so this is fine to me